### PR TITLE
add quantize ansi function

### PR DESCRIPTION
**Pull request by zunda-arrow**
I tested it and so far it works. That was 3am testing though

### CODE REVIEW

File: src/dahlia/utils.py [hunk 1]
```diff
@@ -64,3 +67,224 @@ def _with_marker(marker: str) -> list[Pattern]:
...
+        (128, 128, 128): 90,  # dark gray
+        (255, 0, 0): 91,  # bright red
+        (0, 255, 0): 92,  # bright green
+        (255, 255, 0): 93,  # bright yellow
+        (0, 0, 255): 94,  # bright blue
+        (255, 0, 255): 95,  # bright magenta
+        (0, 255, 255): 96,  # bright cyan
+        (255, 255, 255): 97,  # white
+    },
+}
```
Comment (trag1c) [inline src/dahlia/utils.py]: Feels like these could be moved to `constants.py`?

File: src/dahlia/utils.py [hunk 2]
```diff
@@ -64,3 +67,224 @@ def _with_marker(marker: str) -> list[Pattern]:
...
+
+    if ansi[1] == "2":
+        return _ANSI_24(ansi, old_ansi)
+
+    raise NotImplementedError(
+        "There should never be an ansi that does not follow these rules."
+    )
+
+
+def quantize_ansi(ansi: str, to: Literal[3, 4, 8]) -> str:
```
Comment (trag1c) [inline src/dahlia/utils.py]: Maybe make `to` a keyword-only parameter 🤔 (same for private functions)

File: src/dahlia/utils.py [hunk 3]
```diff
@@ -64,3 +67,224 @@ def _with_marker(marker: str) -> list[Pattern]:
...
+            return _ANSI_4(ansi, old_ansi)
+
+    if ansi[1] == "5":
+        return _ANSI_8(ansi, old_ansi)
+
+    if ansi[1] == "2":
+        return _ANSI_24(ansi, old_ansi)
+
+    raise NotImplementedError(
+        "There should never be an ansi that does not follow these rules."
```
Comment (trag1c) [inline src/dahlia/utils.py]:
```suggestion
        "There should never be an ANSI code that does not follow these rules."
```

File: src/dahlia/utils.py [hunk 4]
```diff
@@ -64,3 +67,224 @@ def _with_marker(marker: str) -> list[Pattern]:
...
+    if len(ansi) < 3:
+        if ansi[0] == 1:
+            color = int(ansi[1])
+        else:
+            color = int(ansi[0])
+
+        if color < 90:
+            return _ANSI_3(ansi, old_ansi)
+        else:
+            return _ANSI_4(ansi, old_ansi)
```
Comment (trag1c) [inline src/dahlia/utils.py:229]: Thoughts on sth like this?
```suggestion
       (_ANSI_3 if color < 90 else _ANSI_4)(ansi, old_ansi)
```

File: src/dahlia/utils.py [hunk 5]
```diff
@@ -64,3 +67,224 @@ def _with_marker(marker: str) -> list[Pattern]:
...
+    (0, 0, 0): 30,  # black
+    (128, 0, 0): 31,  # dark red
+    (0, 128, 0): 32,  # dark green
+    (128, 128, 0): 33,  # dark yellow
+    (0, 0, 128): 34,  # dark blue
+    (128, 0, 128): 35,  # dark magenta
+    (0, 128, 128): 36,  # dark cyan
+    (192, 192, 192): 37,  # light gray
+    (128, 128, 128): 30,  # dark gray
+    # Birght colors are added and linked to the darker ones to improve results.
```
Comment (trag1c) [inline src/dahlia/utils.py]:
```suggestion
    # Bright colors are added and linked to the darker ones to improve results.
```

**Review by trag1c — CHANGES_REQUESTED** [review]
Add a docstring for `quantize_ansi` and include the function in `docs/api/index.rst`

File: src/dahlia/utils.py (hunk 1 above)
Comment (zunda-arrow) [reply src/dahlia/utils.py]: These should be private to the utils files IMO. No way they can ever be used for a different purpose.

File: src/dahlia/utils.py (hunk 4 above)
Comment (zunda-arrow) [reply src/dahlia/utils.py:229]: ugly ngl

File: src/dahlia/utils.py (hunk 1 above)
Comment (trag1c) [reply src/dahlia/utils.py]: same thing applies to almost every other constant :v I'd prefer these to be separated to keep the file a bit more tidy

**Review by trag1c — APPROVED** [review]

**Pull request merged**